### PR TITLE
#2668: Added support for collections and maps with a no-args constructor

### LINKED ISSUE #2668: JSONArray is not being mapped as expected.
I have two fields named the same. both of the type `net.minidev.json.JSONArray`

The generated mapping code is: 
```
JSONArray jSONArray = source.getField();
        if ( jSONArray != null ) {
            ObjectDTO.field( new JSONArray( jSONArray ) );
        }
```
This ctor does not exist in `net.minidev.json.JSONArray`
Is this a bug? which annotation is the best practice in this case? I need to create the exact same JSONArray

### PR DESCRIPTION
**Pull request by Zegveld**
Added support for collections and maps with a no-args constructor
Added a compiler error in case of a collection or map without either a no-arg constructor or a copy constructor.

Fixes #2668

### CODE REVIEW

**Comment by Zegveld** [comment]
This should fix #2668.
It allows support for more collection types. (e.g. other implementations of the java.util collection framework).

**Review by filiphr — CHANGES_REQUESTED** [review]
The functionality looks good @Zegveld. I've left some comments about the implementation, more specifically the 2 new parameters in the `Type`. I don't think that they need to be on the `Type`.

File: processor/src/main/java/org/mapstruct/ap/internal/model/common/Type.java [hunk 1]
```diff
@@ -118,6 +117,8 @@
     private Type boxedEquivalent = null;
 
     private Boolean hasAccessibleConstructor;
+    private boolean hasCopyConstructor;
+    private boolean hasNoArgsConstructor;
```
[the remark this answers is not in the document]
Comment (Zegveld) [reply processor/src/main/java/org/mapstruct/ap/internal/model/common/Type.java]: moved this with https://github.com/mapstruct/mapstruct/pull/2671/commits/b64c708f4e65c7551a44886669f3fd070ea5941e

[the remark this answers is not in the document]
Comment (filiphr) [reply processor/src/main/java/org/mapstruct/ap/internal/model/common/Type.java]: Thanks

**Review by filiphr — APPROVED** [review]
Looks great @Zegveld. I pushed one small change for using a method invocation. We can squash and merge once the CI is done.

If you are merging it, please remove my co-authorship, you did all the work 😉 .

**Pull request merged**